### PR TITLE
Kernel density estimation 1D #15789

### LINKED ISSUE #15789: Kernel density estimation 1D
## Reason
Kernel density is a non-parametric distribution and is frequently used for statistical analysis.

## Design
The KernelDensity1D class is implemented along with tests and documentation.

## Impact
Kernel density will support uncertainty quantification studies through the use of efficient sampling algorithms such as Adaptive Importance Sampling.

### PR DESCRIPTION
**Pull request by somu15**
## Reason
Kernel density is a non-parametric distribution and is frequently used for statistical analysis.

## Design
The KernelDensity1D class is implemented along with tests and documentation.

## Impact
Kernel density will support uncertainty quantification studies through the use of efficient sampling algorithms such as Adaptive Importance Sampling.

closes #15789

### CODE REVIEW

File: modules/stochastic_tools/src/distributions/KernelDensity1D.C [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+InputParameters
+KernelDensity1D::validParams()
+{
+  InputParameters params = Distribution::validParams();
+  MooseEnum bandwidthrule("silverman standarddeviation userdefined");
+  MooseEnum dataformat("vector filename");
+  MooseEnum kernelfunction("gaussian uniform");
+  params.addClassDescription("KernelDensity1D distribution");
+  params.addRequiredParam<MooseEnum>("bandwidthrule", bandwidthrule, "Bandwidth rule for evaluating the bandwith.");
```
Comment (zachmprince) [inline modules/stochastic_tools/src/distributions/KernelDensity1D.C:29]: Change this to `bandwidth_rule`

File: modules/stochastic_tools/src/distributions/KernelDensity1D.C [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  params.addRequiredParam<MooseEnum>("dataformat", dataformat, "Either a data vector or a CSV filename should be provided for building the kernel density.");
+  params.addRequiredParam<MooseEnum>("kernelfunction", kernelfunction, "Kernel function determines the shape of the underlying kernel for the kernel density.");
```
Comment (zachmprince) [inline modules/stochastic_tools/src/distributions/KernelDensity1D.C:31]: Change this to `kernel_function`

**Review by zachmprince — CHANGES_REQUESTED** [review]
Pretty cool functionality. Mainly comments on the interface and testing.

File: modules/stochastic_tools/src/distributions/KernelDensity1D.C [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    std::vector<std::vector<Real>> data1 = reader.getData();
+    _data = data1[0];
+  }
+  Real mu = 0;
+  Real sd = 0;
+  for (unsigned i = 0; i < _data.size(); ++i){
+    mu += _data[i] /  _data.size();
+  }
+  for (unsigned i = 0; i < _data.size(); ++i){
+    sd += std::pow((_data[i] - mu),2) /  _data.size();
```
[the remark this answers is not in the document]
Comment (dschwen) [reply modules/stochastic_tools/src/distributions/KernelDensity1D.C:71]: Also please use `Utility::pow<2>` here. Std pow is super slow and here the exponent is both integer and known at compile time.

**Comment by somu15** [comment]
> Pretty cool functionality. Mainly comments on the interface and testing.

@zachmprince Thank you for your comments. The code looks much better now.

File: modules/stochastic_tools/src/distributions/KernelDensity1D.C [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+InputParameters
+KernelDensity1D::validParams()
+{
+  InputParameters params = Distribution::validParams();
+  MooseEnum bandwidthrule("silverman standarddeviation userdefined");
+  MooseEnum dataformat("vector filename");
+  MooseEnum kernelfunction("gaussian uniform");
+  params.addClassDescription("KernelDensity1D distribution");
+  params.addRequiredParam<MooseEnum>("bandwidthrule", bandwidthrule, "Bandwidth rule for evaluating the bandwith.");
+  params.addRequiredParam<MooseEnum>("dataformat", dataformat, "Either a data vector or a CSV filename should be provided for building the kernel density.");
```
[the remark this answers is not in the document]
Comment (somu15) [reply modules/stochastic_tools/src/distributions/KernelDensity1D.C]: I made this modification.

File: modules/stochastic_tools/src/distributions/KernelDensity1D.C [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+  params.addRequiredParam<MooseEnum>("kernelfunction", kernelfunction, "Kernel function determines the shape of the underlying kernel for the kernel density.");
+  params.addRangeCheckedParam<Real>(
+      "bandwidth", 1.0, "bandwidth > 0", "Bandwidth controls the smoothness of the kernel density.");
+  params.addParam<std::vector<Real>>("data","The data vector.");
```
[the remark this answers is not in the document]
Comment (somu15) [reply modules/stochastic_tools/src/distributions/KernelDensity1D.C]: I added three more tests for data, user defined bandwidth, and standard deviation bandwidth.

File: modules/stochastic_tools/src/distributions/KernelDensity1D.C [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+  if (_dataformat == "filename" && !_file_name_set){
+    paramError("file_name",
+               "A CSV file name needs to be provided if data format equals file_name is chosen.");
+  }
+  if (_dataformat == "filename"){
+    MooseUtils::DelimitedFileReader reader(_file_name);
+    reader.setHeaderFlag(MooseUtils::DelimitedFileReader::HeaderFlag::OFF);
+    reader.read();
+    std::vector<std::vector<Real>> data1 = reader.getData();
+    _data = data1[0];
```
[the remark this answers is not in the document]
Comment (somu15) [reply modules/stochastic_tools/src/distributions/KernelDensity1D.C:64]: Can you please tell me how you want this done? Maybe refer me to another .C file?

File: modules/stochastic_tools/src/distributions/KernelDensity1D.C [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+  if (kernelfunction == "gaussian"){
+    for (unsigned i = 0; i < data.size(); ++i){
+      value += 1/(data.size() * bandwidth) * Normal::pdf(((x - data[i]) / bandwidth), 0.0, 1.0);
+    }
+  } else if (kernelfunction == "uniform"){
+    for (unsigned i = 0; i < data.size(); ++i){
+      value += 1/(data.size() * bandwidth) * Uniform::pdf(((x - data[i]) / bandwidth), -1.0, 1.0);
+    }
+  }
+  return value;
```
[the remark this answers is not in the document]
Comment (somu15) [reply modules/stochastic_tools/src/distributions/KernelDensity1D.C:110]: I added this to the code.

File: modules/stochastic_tools/src/distributions/KernelDensity1D.C [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+  return params;
+}
+
+KernelDensity1D::KernelDensity1D(const InputParameters & parameters)
+  : Distribution(parameters),
+    _bandwidthrule(getParam<MooseEnum>("bandwidthrule")),
+    _dataformat(getParam<MooseEnum>("dataformat")),
+    _kernelfunction(getParam<MooseEnum>("kernelfunction")),
+    _bandwidth(getParam<Real>("bandwidth")),
+    _data_set(parameters.isParamSetByUser("data")),
```
[the remark this answers is not in the document]
Comment (somu15) [reply modules/stochastic_tools/src/distributions/KernelDensity1D.C]: I added this.

File: modules/stochastic_tools/include/distributions/KernelDensity1D.h [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+  /// The bandwith parameter which controls the smoothness of the distribution
+  Real _bandwidth;
+
+  /// data and file_name help get the vector data for building the kernel density
+  bool _data_set;
+  bool _file_name_set;
+
+  std::vector<Real> _data;
+  const FileName _file_name;
```
[the remark this answers is not in the document]
Comment (somu15) [reply modules/stochastic_tools/include/distributions/KernelDensity1D.h:52]: I added this.

File: modules/stochastic_tools/src/distributions/KernelDensity1D.C (hunk 6 above)
[the remark this answers is not in the document]
Comment (zachmprince) [reply modules/stochastic_tools/src/distributions/KernelDensity1D.C:64]: I'm thinking that you can have a parameter like
```
params.addParam<std::string>("file_column_name", "Name of column in csv file to use, by default first column is used.");
```

Then here you can do:
```
MooseUtils::DelimitedFileReader reader(_file_name);
reader.setHeaderFlag(MooseUtils::DelimitedFileReader::HeaderFlag::OFF);
reader.read();
if (isParamValid("file_column_name")
   _data = reader.getData(getParam<std::string>("file_column_name"));
else
   _data = reader.getData(0);
```

File: modules/stochastic_tools/src/distributions/KernelDensity1D.C [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+  {
+    paramError("data", "data and file_name both cannot be set at the same time.");
+  }
+  if (file_name_set)
+  {
+    MooseUtils::DelimitedFileReader reader(_file_name);
+    reader.setHeaderFlag(MooseUtils::DelimitedFileReader::HeaderFlag::OFF);
+    reader.read();
+    std::vector<std::vector<Real>> data1 = reader.getData();
+    _data = data1[0];
```
Comment (zachmprince) [inline modules/stochastic_tools/src/distributions/KernelDensity1D.C:64]: This is how I would change this code:
```suggestion
  params.addParam<FileName>("file_name", "Name of the CSV file.");
  params.addParam<std::string>("file_column_name", "Name of column in csv file to use, by default first column is used.");
  return params;
}

KernelDensity1D::KernelDensity1D(const InputParameters & parameters)
  : Distribution(parameters),
    _bandwidth_rule(getParam<MooseEnum>("bandwidth_rule")),
    _kernel_function(getParam<MooseEnum>("kernel_function")),
    _bandwidth(getParam<Real>("bandwidth"))
{
  if (isParamValid("data") && isParamValid("file_name"))
    paramError("data", "data and file_name both cannot be set at the same time.");
  else if (isParamValid("file_name"))
  {
    MooseUtils::DelimitedFileReader reader(getParam<FileName>("file_name"));
    reader.read();
    if (isParamValid("file_column_name")
       _data = reader.getData(getParam<std::string>("file_column_name"));
    else
       _data = reader.getData(0);
  }
  else if (isParamValid("data"))
    _data = getParam<std::vector<Real>>("data");
  else
    mooseError("Either 'data' or 'file_name' parameters must be specified to represent input data");
```

I might have made a mistake here, so don't commit this directly. You also won't need the `_file_name` variable if you use this logic.

File: modules/stochastic_tools/test/tests/distributions/tests [hunk 11]
```diff
@@ -22,6 +22,61 @@
...
 
+    [kernel1d_gaussian]
+      type = CSVDiff
+      input = kernel_density_1d_gaussian.i
+      csvdiff = 'kernel_density_1d_gaussian_out.csv'
+      allow_test_objects = true
+      abs_zero = 5e-7
+      rel_err = 1e-3
+
+      detail = 'Kernel Density 1D with a Gaussian kernel and data file as input'
```
Comment (zachmprince) [inline modules/stochastic_tools/test/tests/distributions/tests:33]:
```suggestion
      detail = 'Kernel Density 1D with a Gaussian kernel and data file as input, '
```

File: modules/stochastic_tools/test/tests/distributions/tests [hunk 12]
```diff
@@ -22,6 +22,61 @@
...
+
+    [kernel1d_uniform]
+      type = CSVDiff
+      input = kernel_density_1d_uniform.i
+      csvdiff = 'kernel_density_1d_uniform_out.csv'
+      allow_test_objects = true
+      abs_zero = 5e-7
+      rel_err = 1e-3
+
+      detail = 'Kernel Density 1D with a Uniform kernel and data file as input'
```
Comment (zachmprince) [inline modules/stochastic_tools/test/tests/distributions/tests:44]:
```suggestion
      detail = 'Kernel Density 1D with a Uniform kernel and data file as input, '
```

File: modules/stochastic_tools/test/tests/distributions/tests [hunk 13]
```diff
@@ -22,6 +22,61 @@
...
+
+    [kernel1d_datvec]
+      type = CSVDiff
+      input = kernel_density_1d_datvec.i
+      csvdiff = 'kernel_density_1d_datvec_out.csv'
+      allow_test_objects = true
+      abs_zero = 5e-7
+      rel_err = 1e-3
+
+      detail = 'Kernel Density 1D with a Gaussian kernel and data vector as input'
```
Comment (zachmprince) [inline modules/stochastic_tools/test/tests/distributions/tests:55]:
```suggestion
      detail = 'Kernel Density 1D with a Gaussian kernel and data vector as input, '
```

File: modules/stochastic_tools/test/tests/distributions/tests [hunk 14]
```diff
@@ -22,6 +22,61 @@
...
+
+    [kernel1d_userbw]
+      type = CSVDiff
+      input = kernel_density_1d_userbw.i
+      csvdiff = 'kernel_density_1d_userbw_out.csv'
+      allow_test_objects = true
+      abs_zero = 1e-3
+      rel_err = 1e-3
+
+      detail = 'Kernel Density 1D with a Gaussian kernel and user defined bandwidth'
```
Comment (zachmprince) [inline modules/stochastic_tools/test/tests/distributions/tests:66]:
```suggestion
      detail = 'Kernel Density 1D with a Gaussian kernel and user defined bandwidth, '
```

File: modules/stochastic_tools/test/tests/distributions/tests [hunk 15]
```diff
@@ -22,6 +22,61 @@
...
+
+    [kernel1d_sdbw]
+      type = CSVDiff
+      input = kernel_density_1d_sdbw.i
+      csvdiff = 'kernel_density_1d_sdbw_out.csv'
+      allow_test_objects = true
+      abs_zero = 1e-3
+      rel_err = 1e-3
+
+      detail = 'Kernel Density 1D with a Gaussian kernel and standard deviation as bandwidth'
```
Comment (zachmprince) [inline modules/stochastic_tools/test/tests/distributions/tests:77]:
```suggestion
      detail = 'Kernel Density 1D with a Gaussian kernel and standard deviation as bandwidth, '
```

**Review by zachmprince — CHANGES_REQUESTED** [review]
Just a few more pedantic comments. Let me know if I need to give more explanation.

File: modules/stochastic_tools/src/distributions/KernelDensity1D.C (hunk 10 above)
Comment (somu15) [reply modules/stochastic_tools/src/distributions/KernelDensity1D.C]: @zachmprince This works perfectly! All the tests pass. Thanks.

File: modules/stochastic_tools/include/distributions/KernelDensity1D.h [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+  const MooseEnum & _bandwidth_rule;
+  /// kernel_function helps the user select between the different kernel functions that are available
+  const MooseEnum & _kernel_function;
+
+  /// The bandwith parameter which controls the smoothness of the distribution
+  Real _bandwidth;
+
+  /// data and file_name help get the vector data for building the kernel density
+  std::vector<Real> _data;
+  const FileName _file_name;
```
Comment (zachmprince) [inline modules/stochastic_tools/include/distributions/KernelDensity1D.h:52]: You can remove this now.

**Review by zachmprince — CHANGES_REQUESTED** [review]
One last thing, then we're good.

File: modules/stochastic_tools/include/distributions/KernelDensity1D.h (hunk 16 above)
Comment (somu15) [reply modules/stochastic_tools/include/distributions/KernelDensity1D.h]: Thanks pointing this!

**Review by zachmprince — APPROVED** [review]

**Pull request merged**